### PR TITLE
Rework web to fix https not working some some linux machines.

### PR DESCRIPTION
**Pull request by ollydev**
Issue:
------
Synapse & TFPHTTPClient currently access violation on some Linux machines when accessing a https page. Unsure to the exact cause but it's deep in the sockets & OpenSSL files, something we can't really patch out.

Fix:
----
Use web files from FPC trunk! It fixes issue above and brings a lot of upgrades to TFPHTTPClient. The files (six of them) are located [here](https://github.com/ollydev/Simba/tree/9d29cc24a946bf2ec9134da49255bbc6548790e6/Units/Misc/Web/fpc) and can be deleted and uses updated once FPC 3.2 releases. 

Changes:
----------

**Archive additions:**
- Derived from [TSimbaArchiveExtractor ](https://github.com/ollydev/Simba/blob/web-rework/Units/Misc/Archive/simba.archive.pas#L23)Simba can now extract `zip, tar, tar.bz2 and tar.gz`

**Added TSimbaHTTPClient:**
- [TSimbaHTTPClient ](https://github.com/ollydev/Simba/blob/web-rework/Units/Misc/Web/simba.httpclient.pas#L72)is currently only used internally, It's simple and powerful HTTPClient which can eventually be exported to scripts.
- Has download and extract progress callbacks.
- Simple downloading and extracting of the supported archive types (above).
- And everything else you would expect from a http client should do (requests).

**Added asynchronous http requests:**
- [TSimbaHTTPRequest](https://github.com/ollydev/Simba/blob/web-rework/Units/Misc/Web/simba.httpclient_async.pas#L12), [TSimbaHTTPRequestZIP](https://github.com/ollydev/Simba/blob/web-rework/Units/Misc/Web/simba.httpclient_async.pas#L32)
 ```pascal
Request := TSimbaHTTPRequest.Create('https://villavu.com/forum/forum.php');
while Request.Running() do
  Sleep(50);
  
WriteLn(Request.Contents);
WriteLn(Request.ResponseCode);
WriteLn(Request.Exception <> nil); 
  ```

**Environment changes:**
 - Added [TSimbaEnvironment](https://github.com/ollydev/Simba/blob/9d29cc24a946bf2ec9134da49255bbc6548790e6/Projects/Simba/simba.environment.pas) class, which manages simba's default environment (IncludePath etc).
 - Packaged OpenSSL for Linux, now https works by default.

In a following PR I will update packages to utilize this PR.

### CODE REVIEW

**Comment by MerlijnWajer** [comment]
Hm... So you want to keep copied of FPC files in our repo? Otherwise this looks cool.

**Comment by ollydev** [comment]
Temporarily until FPC 3.2 is released yes. It fixes the https bug and includes lots of upgrades so well worth. Once FPC 3.2 is released we update our uses `simba.openssl` >> `openssl` etc then delete the files.

**Comment by ollydev** [comment]
Sneaky force push to remove some commits when I was testing, I guess github doesn't allow you to do it sneakily anymore >.>

But yeah, turns out FPC has it's own resource system where it'll handle it on build. diff hell is no longer. 

OpenSSL libs are stored [here](https://github.com/MerlijnWajer/Simba/tree/fb4cac2e426b9128046044ca7e06d7be9ff52731/Units/Misc/Web/openssl) and automatically added to the Simba executable when built and extracted on runtime when needed.

File: Projects/Simba/newsimbasettings.pas [hunk 1]
```diff
@@ -179,21 +179,17 @@     TTraySection = class(TSection)
     end;
 
     TSourceEditorSection = class(TSection)
-      DefScriptPath: TFileSetting;
+      DefScriptPath: TStringSetting;
       LazColors: TBooleanSetting;
       CaretPastEOL: TBooleanSetting;
       Font: TFontSetting;
-      HighlighterPath: TFileSetting;                            
+      HighlighterPath: TStringSetting;
```
Comment (MerlijnWajer) [inline Projects/Simba/newsimbasettings.pas:186]: Does this belong in the same PR?

Comment (ollydev) [reply Projects/Simba/newsimbasettings.pas:186]: The TLDR is that I rewrote how the environment is handled so installing openssl libs would be easier.
It was a relatively [small change](https://github.com/MerlijnWajer/Simba/blob/e146e1cc107f380ca7d2441207560f5433f7c98d/Projects/Simba/simba.environment.pas) so it's much easier to keep it in one pull request than queuing and syncing pull requests to you.

If curious, no outside functionality changed here, only internally.

**Comment by MerlijnWajer** [comment]
Boom.

**Pull request merged**